ParameterAlgo: add manual conversion from BoolVectorData to AtArray.

mattigruener opened the pull request:
Hey John,

I assume you had something like this in mind regarding the conversion from BoolVectorData to AtArray? The corresponding function that deals with motion samples should already work if I interpret the despatchTypedData calls correctly?

Thanks,

Matti.

mattigruener commented:
I don't think I did interpret it correctly and a change is still necessary. I do find the despatchTypedData bits to be a bit confusing to be honest.

johnhaddon commented on contrib/IECoreArnold/src/IECoreArnold/ParameterAlgo.cpp:
```diff
@@ -441,6 +441,18 @@ AtArray *dataToArray( const IECore::Data *data, int aiType )
+	// convert to an AtArray here.
+	else if( aiType == AI_TYPE_BOOLEAN )
+	{
+		const BoolVectorData* booleans = static_cast<const BoolVectorData *>( data );
```
This is the gist of what I was getting at, but let's just make a few small tweaks :

- Call `readable()` once outside the loop rather than twice for every element. The cost is relatively small, but I've measured it in the past and it is worth avoiding.
- Call size() once rather than once per iteration, for the same reason.
- Use `[]` rather than `at()`. We don't need the bounds checking that `at()` will do, since we know we're iterating over the right range. This removes overhead, but it also the intention clearer.
- Use `size_t` (or `vector<bool>::size_type` if you want to out-pedant me) for the loop index.

Here's an example of some of that :

https://github.com/ImageEngine/cortex/blob/master/contrib/IECoreArnold/src/IECoreArnold/MeshAlgo.cpp#L207

johnhaddon commented:
Just saw your second comment and am a bit confused as I think what you have should be working - is it not? Maybe add a test case to put our minds at rest?

mattigruener commented:
I didn't mean to confuse you - take my comment with a grain of salt because as I said, I don't have a thorough understanding of what the despatchTypedData is doing yet. I will look into addressing your feedback and adding a test case so that we can sleep at night. Thanks! :)

mattigruener commented:
To be clear - when I said a change was still necessary, I meant the function that deals with motion samples (https://github.com/mattigruener/cortex/blob/086bba2f7b80f72766abbb31a8d7bec5ca44cad0/contrib/IECoreArnold/src/IECoreArnold/ParameterAlgo.cpp#L462). I did test my changes in this PR and that worked fine, but I haven't built a test for triggering the code that deals with the samples/keys yet.

johnhaddon commented:
I see - got it. In theory you would need to make a similar specialised version for the motion samples case, but I reckon we should just ignore that - I can't think of any likely scenario where it'd be necessary to interpolate between bools. Let's deal with it if and when it comes up.

mattigruener commented:
Addressed your feedback. Let me know if there's anything else you want changed.

The test in the 2nd commit fails without the conversion introduced by the 1st. I noticed that some other tests fail, but they do so independent of my addition. 

Thanks :)

johnhaddon commented on contrib/IECoreArnold/src/IECoreArnold/ParameterAlgo.cpp:
```diff
@@ -441,6 +441,19 @@ AtArray *dataToArray( const IECore::Data *data, int aiType )
+	// convert to an AtArray here.
+	else if( aiType == AI_TYPE_BOOLEAN )
+	{
+		vector<bool> booleans = static_cast<const BoolVectorData *>( data )->readable();
```
This should be a reference rather than a copy.

johnhaddon commented on contrib/IECoreArnold/test/IECoreArnold/MeshTest.py:
```diff
@@ -243,6 +243,22 @@ def testPointTypePrimitiveVariables( self ) :
+		m = IECore.MeshPrimitive.createPlane( IECore.Box2f( IECore.V2f( -1 ), IECore.V2f( 1 ) ) )
+		m["myBoolPrimVar"] = IECore.PrimitiveVariable(
+			IECore.PrimitiveVariable.Interpolation.Vertex,
+			IECore.BoolVectorData( [ True, False ] )
```
A vertex primvar for a plane should have 4 values shouldn't it?

mattigruener commented:
Updated. Those made a lot of sense of course :)

johnhaddon commented:
Nice one - thanks Matti!
